service/parserprovider/expand_test.go

```diff
@@ -0,0 +1,119 @@
+// limitations under the License.
+
+package parserprovider
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"go.opentelemetry.io/collector/config"
+)
+
+func TestExpand(t *testing.T) {
+	var testCases = []struct {
+		name string // test case name (also file name containing config yaml)
+	}{
+		{name: "expand-with-no-env.yaml"},
+		{name: "expand-with-partial-env.yaml"},
+		{name: "expand-with-all-env.yaml"},
+	}
+
+	const valueExtra = "some string"
+	const valueExtraMapValue = "some map value"
+	const valueExtraListElement = "some list value"
+	assert.NoError(t, os.Setenv("EXTRA", valueExtra))
+	assert.NoError(t, os.Setenv("EXTRA_MAP_VALUE_1", valueExtraMapValue+"_1"))
+	assert.NoError(t, os.Setenv("EXTRA_MAP_VALUE_2", valueExtraMapValue+"_2"))
+	assert.NoError(t, os.Setenv("EXTRA_LIST_VALUE_1", valueExtraListElement+"_1"))
+	assert.NoError(t, os.Setenv("EXTRA_LIST_VALUE_2", valueExtraListElement+"_2"))
+
+	defer func() {
+		assert.NoError(t, os.Unsetenv("EXTRA"))
+		assert.NoError(t, os.Unsetenv("EXTRA_MAP_VALUE"))
+		assert.NoError(t, os.Unsetenv("EXTRA_LIST_VALUE_1"))
+	}()
+
+	expectedCfgMap, errExpected := config.NewMapFromFile(path.Join("testdata", "expand-with-no-env.yaml"))
+	require.NoError(t, errExpected, "Unable to get expected config")
+
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			// Retrieve the config
+			emp := NewExpandMapProvider(&testFileProvider{fileName: path.Join("testdata", test.name)})
+			cfgMap, err := emp.Get(context.Background())
+			require.NoError(t, err, "Unable to get config")
+
+			// Test that expanded configs are the same with the simple config with no env vars.
+			assert.Equal(t, expectedCfgMap.ToStringMap(), cfgMap.ToStringMap())
+		})
+	}
+}
+
+func TestExpand_EscapedEnvVars(t *testing.T) {
+	const receiverExtraMapValue = "some map value"
+	assert.NoError(t, os.Setenv("MAP_VALUE_2", receiverExtraMapValue))
+	defer func() {
+		assert.NoError(t, os.Unsetenv("MAP_VALUE_2"))
+	}()
+
+	// Retrieve the config
+	emp := NewExpandMapProvider(&testFileProvider{fileName: path.Join("testdata", "expand-escaped-env.yaml")})
+	cfgMap, err := emp.Get(context.Background())
+	require.NoError(t, err, "Unable to get config")
+
+	expectedMap := map[string]interface{}{
+		"test_map": map[string]interface{}{
+			// $$ -> escaped $
+			"recv.1": "$MAP_VALUE_1",
+			// $$$ -> escaped $ + substituted env var
+			"recv.2": "$" + receiverExtraMapValue,
+			// $$$$ -> two escaped $
+			"recv.3": "$$MAP_VALUE_3",
+			// escaped $ in the middle
+			"recv.4": "some${MAP_VALUE_4}text",
+			// $$$$ -> two escaped $
+			"recv.5": "${ONE}${TWO}",
+			// trailing escaped $
+			"recv.6": "text$",
+			// escaped $ alone
+			"recv.7": "$",
+		}}
+	assert.Equal(t, expectedMap, cfgMap.ToStringMap())
+}
+
+// TODO: When FileProvider supports passing the file name remove this.
```

Comment:
Please open an issue if one doesnt already exist.

Reply:
Depending on https://github.com/open-telemetry/opentelemetry-collector/pull/4077 which one gets merge first the second will fix this.